langserver/references.go

```diff
@@ -212,6 +212,32 @@ func (h *LangHandler) handleTextDocumentReferences(ctx context.Context, conn JSO
 		t := time.NewTicker(time.Second)
 		defer t.Stop()
 		streamTick = t.C
+		streamPos := 0
+		streamUpdate = func() {
+			mu.Lock()
+			partial := refs
+			mu.Unlock()
+			if len(partial) == streamPos {
```

Comment:
this line means everything in refs so far has already been sent. If refs changes after mu.Unlock but before this line, partial won't change. We need to snapshot the slice (note its just the slice "header", not the actual content of the slice so is small) since reading `refs` outside of the lock is a concurrent read/write (ie undefined behaviour). Holding the lock while we build `patch` is much longer than we need to hold the lock.